### PR TITLE
ARROW-16872: [C++] Fix CSV parser edge case

### PR DESCRIPTION
**Pull request by cyb70289**
Fixes an error when there's no EOL at last line, and last field is consumed by bulk filter.

### CODE REVIEW

**Comment by cyb70289** [comment]
@ursabot please benchmark command=cpp-micro --suite-filter="arrow-csv-*"

File: cpp/src/arrow/csv/parser.cc [hunk 1]
```diff
@@ -296,6 +297,9 @@ class BlockParserImpl {
     if (UseBulkFilter) {
       const char* bulk_end = RunBulkFilter(parsed_writer, data, data_end, bulk_filter);
       if (ARROW_PREDICT_FALSE(bulk_end == nullptr)) {
+        if (is_final) {
+          final_offset = static_cast<int32_t>(data_end - data);
```
Comment (pitrou) [inline cpp/src/arrow/csv/parser.cc:301]: Can this simply be
```suggestion
          data = data_end;
```

Comment (cyb70289) [reply cpp/src/arrow/csv/parser.cc:301]: `data` is used in line https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/parser.cc#L387 if there are fewer rows than expected. Didn't read deep`HandleInvalidRow` code, will updating `data` cause problem?

Comment (pitrou) [reply cpp/src/arrow/csv/parser.cc:301]: We could perhaps add a test for this fringe situation...

File: cpp/src/arrow/csv/parser.cc [hunk 2]
```diff
@@ -337,6 +340,9 @@ class BlockParserImpl {
     if (UseBulkFilter) {
       const char* bulk_end = RunBulkFilter(parsed_writer, data, data_end, bulk_filter);
       if (ARROW_PREDICT_FALSE(bulk_end == nullptr)) {
+        if (is_final) {
+          data = data_end;
+        }
```
Comment (cyb70289) [inline cpp/src/arrow/csv/parser.cc:345]: Is it an error if this happens in a quoted field?

File: cpp/src/arrow/csv/parser.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/parser.cc]: Done

**Comment by cyb70289** [comment]
CI error is strange: https://github.com/apache/arrow/runs/7125162553?check_suite_focus=true#step:6:3704
```
[ RUN      ] BlockParser.FinalTruncatedBulkFilterNoEol
/arrow/cpp/src/arrow/csv/parser_test.cc:423: Failure
Expected equality of these values:
  ("Invalid: CSV parse error: Expected 2 columns, got 1: 87654321")
    Which is: "Invalid: CSV parse error: Expected 2 columns, got 1: 87654321"
  _st.ToString()
    Which is: "Invalid: CSV parse error: Expected 2 columns, got 1: 87654321\n/arrow/cpp/src/arrow/csv/parser.cc:464  (ParseLine<SpecializedOptions, true>(values_writer, parsed_writer, data, data_end, is_final, &line_end, bulk_filter))\n/arrow/cpp/src/arrow/csv/parser.cc:566  ParseChunk<SpecializedOptions>( &values_writer, &parsed_writer, data, data_end, is_final, rows_in_chunk, &data, &finished_parsing, bulk_filter)"
With diff:
@@ +1,3 @@
 Invalid: CSV parse error: Expected 2 columns, got 1: 87654321
+/arrow/cpp/src/arrow/csv/parser.cc:464  (ParseLine<SpecializedOptions, true>(values_writer, parsed_writer, data, data_end, is_final, &line_end, bulk_filter))
+/arrow/cpp/src/arrow/csv/parser.cc:566  ParseChunk<SpecializedOptions>( &values_writer, &parsed_writer, data, data_end, is_final, rows_in_chunk, &data, &finished_parsing, bulk_filter)
[  FAILED  ] BlockParser.FinalTruncatedBulkFilterNoEol (0 ms)
```
Looks like some gtest debug messages are appended to the status string.

**Comment by pitrou** [comment]
Perhaps ARROW_EXTRA_ERROR_CONTEXT is enabled?

File: cpp/src/arrow/csv/parser.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/csv/parser.cc:345]: Hmm... I don't *think* so.

**Review by pitrou — APPROVED** [review]
+1, thanks for the fix @cyb70289

**Pull request merged**